test_version.sh

```diff
@@ -2,37 +2,95 @@
 
 set -e
 
-test_scala_version() {
-  SCALA_VERSION=$1
-
-  SCALA_VERSION_SHAS=''
-  SCALA_VERSION_SHAS+='"scala_compiler": "'$2'",'
-  SCALA_VERSION_SHAS+='"scala_library": "'$3'",'
-  SCALA_VERSION_SHAS+='"scala_reflect": "'$4'"'
+scala_2_11_version="2.11.12"
+scala_2_11_shas='\
+  "scala_compiler": "3e892546b72ab547cb77de4d840bcfd05c853e73390fed7370a8f19acb0735a0", \
+  "scala_library": "0b3d6fd42958ee98715ba2ec5fe221f4ca1e694d7c981b0ae0cd68e97baf6dce", \
+  "scala_reflect": "6ba385b450a6311a15c918cf8688b9af9327c6104f0ecbd35933cfcd3095fe04" \
+'
+scala_2_12_version="2.12.10"
+scala_2_12_shas='\
+  "scala_compiler": "cedc3b9c39d215a9a3ffc0cc75a1d784b51e9edc7f13051a1b4ad5ae22cfbc0c", \
+  "scala_library": "0a57044d10895f8d3dd66ad4286891f607169d948845ac51e17b4c1cf0ab569d", \
+  "scala_reflect": "56b609e1bab9144fb51525bfa01ccd72028154fc40a58685a1e9adcbe7835730" \
+'
+
+SCALA_VERSION_DEFAULT=$scala_2_11_version
+SCALA_VERSION_SHAS_DEFAULT=$scala_2_11_shas
+TWITTER_SCROOGE_VERSION_SHAS_DEFAULT=''
+TWITTER_SCROOGE_EXTRA_IMPORTS_DEFAULT=''
+TWITTER_SCROOGE_BINDINGS_DEFAULT='twitter_scrooge(scala_version)'
+
+run_in_test_repo() {
+  SCALA_VERSION=${SCALA_VERSION:-$SCALA_VERSION_DEFAULT}
+  SCALA_VERSION_SHAS=${SCALA_VERSION_SHAS:-$SCALA_VERSION_SHAS_DEFAULT}
+  TWITTER_SCROOGE_BINDINGS=${TWITTER_SCROOGE_BINDINGS:-$TWITTER_SCROOGE_BINDINGS_DEFAULT}
+  TWITTER_SCROOGE_EXTRA_IMPORTS=${TWITTER_SCROOGE_EXTRA_IMPORTS:-$TWITTER_SCROOGE_EXTRA_IMPORTS_DEFAULT}
+  TWITTER_SCROOGE_VERSION_SHAS=${TWITTER_SCROOGE_VERSION_SHAS:-$TWITTER_SCROOGE_VERSION_SHAS_DEFAULT}
+
+  test_command=$1
+  test_dir_prefix=$2
 
   cd "${dir}"/test_version
 
   timestamp=$(date +%s)
 
-  NEW_TEST_DIR="test_${SCALA_VERSION}_${timestamp}"
+  NEW_TEST_DIR="test_${test_dir_prefix}_${timestamp}"
 
   cp -r version_specific_tests_dir/ $NEW_TEST_DIR
 
   sed \
       -e "s/\${scala_version}/$SCALA_VERSION/" \
-      -e "s/\${scala_version_shas}/$SCALA_VERSION_SHAS/" \
+      -e "s%\${scala_version_shas}%$SCALA_VERSION_SHAS%" \
+      -e "s%\${twitter_scrooge_bindings}%$TWITTER_SCROOGE_BINDINGS%" \
```

Comment:
I agree, this mechanism was not particularly easy to follow, and. this file has now grown to be an uncomfortable amount of bash. 

I'm wondering how much coordination this would need with the toolchains work, if that's what we want to use for customization in the future.

Reply:
Regardless, I'll open a ticket about it, although I probably won't be able to look into it in the near future, we could start discussing.

Reply:
Opened https://github.com/bazelbuild/rules_scala/issues/1057